### PR TITLE
Refactor CVR Readers to inherit from same base class

### PR DESCRIPTION
**Pull request by artoonie**
Refactor readers in preparation for Issue 60.

Motivation:
Currently, only one place in the code needed to branch based on the `Provider`: when tabulating.
To accomplish Issue 60, we'll need each reader to also be able to provide a list of Candidates.
This refactor will make it easier to accomplish.

Further: this seems to help with #621 as well: before, only one reader could support tabulating by precinct. With this refactor, it makes it a little cleaner to implement tabulate-by-precinct across readers, or at least have some default implementation. That is not done yet, but it will help.

### CODE REVIEW

**Review by tarheel — APPROVED** [review]
Nice!

**Review by HEdingfield — CHANGES_REQUESTED** [review]
Putting a hold on this until we get the PR it depends on merged first. I want to see how the dust settles after that merge, since some of the comments I left in [this review](https://github.com/BrightSpots/rcv/pull/649#pullrequestreview-1435460261) might apply here as well.

**Comment by artoonie** [comment]
Squash merge bit me again!

How committed are we to squash-merges? They make chained PRs very painful, and they've already led to lost changes. I personally prefer standard merges -- it mirrors the "real" history better -- but understand that squash-merges make cleaner git logs at the expense of some issues like this.

See also:
https://abe-winter.github.io/2019/09/01/squanch-merge.html
https://stackoverflow.com/questions/54636294/github-pr-squash-causes-erroneous-merge-conflicts

**Comment by artoonie** [comment]
@HEdingfield this is ready for review

**Comment by tarheel** [comment]
> How committed are we to squash-merges? They make chained PRs very painful, and they've already led to lost changes. I personally prefer standard merges -- it mirrors the "real" history better -- but understand that squash-merges make cleaner git logs at the expense of some issues like this.

Not my area of expertise, but I personally have a pretty strong preference for squashing because, as you note, it prevents the messiness of each PR from leaking into the commit history of `main`. I understand that it can make it trickier to manage stacked PRs, though. I believe there are a number of tools out there that are designed to mitigate this exact problem. Maybe we should look into them for this project...

I'll also note that this problem just hasn't really come up much for us in the past because we've rarely had multiple PRs in flight that were interdependent in a complex way.

**Comment by tarheel** [comment]
@HEdingfield this still seems fine to me. I'll let you approve.

**Comment by HEdingfield** [comment]
> > How committed are we to squash-merges? They make chained PRs very painful, and they've already led to lost changes. I personally prefer standard merges -- it mirrors the "real" history better -- but understand that squash-merges make cleaner git logs at the expense of some issues like this.
> 
> Not my area of expertise, but I personally have a pretty strong preference for squashing because, as you note, it prevents the messiness of each PR from leaking into the commit history of `main`. I understand that it can make it trickier to manage stacked PRs, though. I believe there are a number of tools out there that are designed to mitigate this exact problem. Maybe we should look into them for this project...
> 
> I'll also note that this problem just hasn't really come up much for us in the past because we've rarely had multiple PRs in flight that were interdependent in a complex way.

I agree with @tarheel and, as I mentioned in another thread, this is what I use at work so I'm pretty used to it at this point. For chained PRs (e.g. PR1 based on dev, PR2 based on PR1), we do this:

* Merge PR1 into dev
* Merge dev back into PR2
* Merge PR2 into dev

We typically don't have any issues by following this consistently. I'm not exactly sure what happened in this case to cause lost changes, but it's definitely something we can keep an eye on to try to get to the bottom of it if it happens again.

File: src/main/java/network/brightspots/rcv/CsvCvrReader.java
Comment (HEdingfield) [inline src/main/java/network/brightspots/rcv/CsvCvrReader.java:3]: Sorry, this may seem like a dumb change, but it happened due to the way `BrightSpotsCopyrightProfile.xml` is set up. I have an open issue to update this, and will update all other files at the same time.

**Review by HEdingfield — APPROVED** [review]
I found a few little things to fix, and then ran IntelliJ auto-format over the modified files (hence so many indentation / linting changes). This looks good to me now!

**Comment by HEdingfield** [comment]
@artoonie I'll go ahead and merge this in now, but please do check over the changes I made / the question I asked if you get a chance so we can iterate further if needed. Great work on this!

File: src/main/java/network/brightspots/rcv/CsvCvrReader.java
Comment (artoonie) [reply src/main/java/network/brightspots/rcv/CsvCvrReader.java]: Good point -- this argument is kind of ugly. I cleaned it up here: https://github.com/BrightSpots/rcv/pull/683/files

**Comment by HEdingfield** [comment]
> The issue will arise if you:
> 
> 1. Build chain PR1 < PR2
> 2. PR1 gets a code review; commits added
> 3. Merge PR1 into PR2
> 4. Merge PR1 into dev
> 5. Merge dev into PR2
> 
> Step 3 is the culprit: the squash-merge loses track of the merge-commit. I don't know of any smarter tools to deal with this, happy to explore.

Interesting... so where exactly is data lost in that sequence? Are you saying step 5 doesn't pick up all the changes you'd expect from dev?

I'd think that it would throw the squashed commit at the top of the commit list, basically overwriting / replicating all the other ones upstream. Then when you approve PR2 and squash merge back into dev, it should be one nice, clean commit.

> Another option is to allow non-squash commits for non-chained PRs, which avoids messy commit history most of the time.

I think this is exactly what we're trying to avoid in the first place though; in my experience, non-chained PRs can easily be just as messy, commit-wise, as a chained one. Just thinking of all the times I've seen "lint fix", "responding to PR comments", "more linting", "more PR comments", "fuck this linter", etc. etc. all in a row. :/

**Comment by artoonie** [comment]
No data is "lost" by git, but you will have to solve so many merge conflicts that it's easy for a change to get lost -- which has happened a few times already. It's so bad it's even in the git man pages: https://github.com/git/git/blob/5065ce412ef083a02288c1972ea3d07423cace0e/Documentation/gitfaq.txt#L244-L275

I think my example before wasn't actually what is happening. It's not the code review that causes issues, but merge conflict resolution. As an example, let's say you have two features, `feature/branch-1` is a single branch, and `feature/branch-2-part-x` is a PR chain. If `feature/branch-1` merges first, and has a conflict with `feature/branch-2-part-1`, then the squash merge will cause that merge conflict to reappear in every future PR for `feature/branch-2-part-x`.

When using non-squashed merges, we can require clean git history before the merge is allowed: developer can squash all code commits on their end, then add a merge-conflict-resolving commit at the end. This will prevent every future branch in a chain from needing to re-resolve the conflicts.

FWIW, I use `git commit --amend` && `git push --force` to keep git history clean-ish and avoid the 100 `lint fix` commits -- and I think we can require that for non-squashed branches.

**Comment by HEdingfield** [comment]
> I think my example before wasn't actually what is happening. It's not the code review that causes issues, but merge conflict resolution. As an example, let's say you have two features, `feature/branch-1` is a single branch, and `feature/branch-2-part-x` is a PR chain. If `feature/branch-1` merges first, and has a conflict with `feature/branch-2-part-1`, then the squash merge will cause that merge conflict to reappear in every future PR for `feature/branch-2-part-x`.
> 

If we were to require squash commits into `develop`, but allow other types of commits everywhere else, would it address this problem? I'm guessing no, but want to ask anyway.

> When using non-squashed merges, we can require clean git history before the merge is allowed: developer can squash all code commits on their end, then add a merge-conflict-resolving commit at the end. This will prevent every future branch in a chain from needing to re-resolve the conflicts.
> 

Can we actually require a clean history in the GH settings, or do you mean as a guideline to contributors?

Also, would the merge-conflict-resolving commit at the end eventually appear in the `develop` branch history, or is there a way we could clean that up?

> FWIW, I use `git commit --amend` && `git push --force` to keep git history clean-ish and avoid the 100 `lint fix` commits -- and I think we can require that for non-squashed branches.

That's awesome, and similar to what I do in Mercurial. However, I don't think there's any way to actually enforce this, unless I'm mistaken? We've had a number of past contributors who don't do it this way and things get messy.

My overall goal is this: `develop` should have a very clean commit history, where each commit equates to resolving a specific issue, with no other junk in between. If we can accomplish that (ideally in a GH-enforceable way), I'm not opposed to making other tweaks to our current flow.

**Comment by artoonie** [comment]
1. Would requiring squash into develop solve the problem? **No**
2. Can github enforce clean history, or is it just a guideline? **Just a guideline**
3. Would the merge-conflict-resolve appear in `develop`'s history? **Yes**

I have another idea for how to resolve this:
If a PR chain is entirely approved: `develop < PR1 < PR2`, this problem would also be resolved by merging `PR1 < PR2` then merging `Develop < PR1`. That will create one mega-commit into develop, and create a too-large pull request, but if each member of the chain was individually approved, this could work fine. Resolves the issue of keeping git history clean -- this time with even fewer commits! -- with the downside that chains have to live longer to wait for the entire chain to be approved.

**Pull request merged**
